### ISSUE TITLE
Plugins in luakit-plugins repo are broken

### ISSUE THREAD

**Issue by schvabodka-man** [luakit/luakit#457]
And i've fixed 3 of them(still can't fix user agent changer) and now they work nice with new api. It would be good if you'll merge this
https://github.com/luakit/luakit-plugins/pull/14

**Comment by aidanholm** [comment]
Sorry, it seems I wasn't previously watching the luakit plugins repo! Thanks for taking the time to patch those up; I'll take a look at your work now.

**Comment by schvabodka-man** [comment]
It would be nice if you consult me about some things(i don't know a lot about luakit architecture and api) and i should make user agent switcher work. 
1) How do i access some file? In fact file from plugins directory
2) What is variable that store browser useragent and how to i get to this variable? Useragent switcher are refering to globals.useragent but it seems like it's fucked(i'm getting error when starting luakit). Is the way browser store useragent changed?

**Comment by aidanholm** [comment]
Sure, happy to help!
1. You can use the `lousy.util.find_*` functions to find the path to a file, and `lousy.load()` will load a file's contents.
2. Useragents are set from `globals.useragent`; if you're getting an error, you might have to add `local globals = require "globals"` first.

**Comment by schvabodka-man** [comment]
Adding require global, getting global.useragent and this occurs

`Lua error: attempt to index global 'string' (a nil value)`

**Comment by aidanholm** [comment]
That shouldn't happen... ever. What's the traceback?

**Comment by schvabodka-man** [comment]
Sorry, mu fuck up, i forgot to import string. Ok. How do i load data from plugin directory? I don't see func for it in lousy.util and find_data doesn't work.

Update:
Okay, i managed to laod it in tricky way with
`util.find_config("plugins/uaswitch/ua_strings.lua")`
And now i'm getting nil error
`[    0.322618] E [core/common/lualib]: Lua error: bad argument #1 to 'rep' (string expected, got nil)`

**Comment by aidanholm** [comment]
You shouldn't have to import `string`; it's a built-in global variable. The only thing I can think of is if you had something like `local string = nil` shadowing the global.

`lousy.util.find_config()` basically concats the path you give it with the config path. So the way you did it should work fine, and is the proper way to do this. What is the traceback for that error?

**Comment by schvabodka-man** [comment]
It seems like it doesn't getting anything from global.useragent or something like that. I'm replaced it by just string and it is starting but i cant fix further errors when running browser and trying to switch user agent. Maybe i should commit some code i've managed to do yet and you take a look because i'm really bad in Lua?

**Comment by aidanholm** [comment]
You could pastebin the output of `git diff` somewhere, that's probably easier

**Comment by schvabodka-man** [comment]
transfer.sh/xiVHY/luakit-diff 
Here it is.  Pretty dirty

**Comment by Plaque-fcc** [comment]
Yeah, they are pretty out of date. Feels like @liuch could help rather more than I could.

On August 21, 2017 7:15:25 PM GMT+03:00, Ivan <notifications@github.com> wrote:
>transfer.sh/xiVHY/luakit-diff 
>Here it is.  Pretty dirty
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/luakit/luakit/issues/457#issuecomment-323788553

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by Plaque-fcc** [comment]
And I'm still interested in letting this live, so feel free to make pull requests to luakit-plugins repo (there should be at least two of them IIRC)...

On August 21, 2017 7:15:25 PM GMT+03:00, Ivan <notifications@github.com> wrote:
>transfer.sh/xiVHY/luakit-diff 
>Here it is.  Pretty dirty
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/luakit/luakit/issues/457#issuecomment-323788553

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by aidanholm** [comment]
@Plaque-fcc There is a PR, thanks to @schvabodka-man :) https://github.com/luakit/luakit-plugins/pull/14

Re `uaswitch.lua`; general comments, haven't ran tests yet:
- Lots of `local foo = foo` lines at the top; they should all be deleted
- Lots of global variables, e.g. `ua_strings = {}`. These should all be made fields of `_M`: `_M.ua_strings = {}`.
- The `ua_strings` file should just be moved into the main file; make it public so that you can just do something like this:
```lua
uaswitch.useragents.firefox = "... useragent here ..."
```

**Comment by Plaque-fcc** [comment]
Aidan, I was distracted, so I thought it was all about the repo I manage. Need to check the situation to stay updated. ^_^

On August 21, 2017 7:38:58 PM GMT+03:00, Aidan Holm <notifications@github.com> wrote:
>@Plaque-fcc There is a PR, thanks to @schvabodka-man :)
>https://github.com/luakit/luakit-plugins/pull/14
>
>Re `uaswitch.lua`; general comments, haven't ran tests yet:
>- Lots of `local foo = foo` lines at the top; they should all be
>deleted
>- Lots of global variables, e.g. `ua_strings = {}`. These should all be
>made fields of `_M`: `_M.ua_strings = {}`.
>- The `ua_strings` file should just be moved into the main file; make
>it public so that you can just do something like this:
>```lua
>uaswitch.useragents.firefox = "... useragent here ..."
>```
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/luakit/luakit/issues/457#issuecomment-323794198

-- 
J.K.

**Comment by schvabodka-man** [comment]
User agents are a little bit out of date. Will it be ok if i would make you list of useragents for most of modern browsers, some exotic ones, consoles, phones/tablests and spiders and you'll update this shit by yourself because i don't have much time at the moment?

**Comment by Plaque-fcc** [comment]
I was quite inconsistent writing this list (thought it can be a little start for later improvements) and most of the plugins. Still didn't rewrite adblock to utilize far more efficient algo and update subscription in background. But I'm gonna commit more.

On August 21, 2017 7:42:47 PM GMT+03:00, Ivan <notifications@github.com> wrote:
>User agents are a little bit out of date. Will it be ok if i would make
>you list of useragents for most of modern browsers, some exotic ones,
>consoles, phones/tablests and spiders and you'll update this shit by
>yourself because i don't have much time at the moment?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/luakit/luakit/issues/457#issuecomment-323795047

-- 
J.K.

**Comment by aidanholm** [comment]
@Plaque-fcc No worries :) Commits are very-much appreciated! Note that I updated the adblock module extensively (lots of efficiency hacks) and it's the only version that will work with the new luakit's APIs, so take a look there before starting from the old version :) Some form of subscriptions updater would be very nice (`:adblock-update`?) 

@schvabodka-man I think there's probably no need to have a massive list of user agents to begin with. Updating the list is a separate task from making the plugin work; I'd be fine with merging a PR that just got the plugin working, even if the UAs are outdated.

And there's no need to finish all the work immediately if you're busy; this can always wait!

**Comment by schvabodka-man** [comment]
Okay, here are small list of user agents for modern browsers. This may come in handy when you'll be pacthing it

```
plugins.uaswitch.ua_strings.fakes = {
   
   ["luakit"]        = "Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/604.1 (KHTML, like Gecko) Version/11.0 Safari/604.1",
   ["firefox"]        = "Mozilla/5.0 (Windows NT 10.0; WOW64; rv:54.0) Gecko/20100101 Firefox/54.0",
   ["chrome"]        = "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36",
   ["chromium"]        = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/58.0.3029.110 Chrome/58.0.3029.110 Safari/537.36",
   ["edge"]        = "Mozilla/5.0 (Windows NT 6.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36 Edge/14.14393",
   ["ie11"]        = "Mozilla/5.0 (Windows NT 10.0; WOW64; Trident/7.0; rv:11.0) like Gecko",
   ["opera"]        = "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36 OPR/46.0.2597.57",
   ["opera-presto"]        = "Opera/9.80 (Windows NT 6.0) Presto/2.12.388 Version/12.14",
   ["opera-mini-j2me"]        = "Opera/9.80 (J2ME/MIDP; Opera Mini/4.2/28.3590; U; en) Presto/2.8.119 Version/11.10",
   ["opera-mini-android"]        = "Opera/9.80 (Android; Opera Mini/24.0.2254/62.178; U; en) Presto/2.12.423 Version/12.16",
   ["okhttp"]        = "okhttp/3.8.0",
   ["netscape"]        = "Mozilla/4.5 (compatible; HTTrack 3.0x; Windows 98)",
   ["netsurf"]        = "NetSurf/3.6 (Linux)",
   ["links"]        = "Links (2.7; Linux 3.10.24+ armv6l; GNU C 4.6.3; dump)",
   ["dillo"]        = "Dillo/3.0.5",
   ["lynx"]        = "Lynx/2.8.8pre.4 libwww-FM/2.14 SSL-MM/1.4.1 GNUTLS/2.12.23",
   ["safari"]        = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_3) AppleWebKit/602.4.8 (KHTML, like Gecko) Version/10.0.3 Safari/602.4.8",
   ["iphone"]        = "Mozilla/5.0 (iPhone; CPU iPhone OS 10_2_1 like Mac OS X) AppleWebKit/602.4.6 (KHTML, like Gecko) Version/10.0 Mobile/14D27 Safari/602.1",
   ["ipad"]        = "Mozilla/5.0 (iPad; CPU OS 9_3_1 like Mac OS X) AppleWebKit/601.1.46 (KHTML, like Gecko) Version/9.0 Mobile/13E238 Safari/601.1",
["android-phone"]        = "Mozilla/5.0 (Linux; Android 6.0.1; Nexus 6P Build/MMB29P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.83 Mobile Safari/537.36",
   ["android-tablet"]        = "Mozilla/5.0 (Linux; Android 7.0; Pixel C Build/NRD90M; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/52.0.2743.98 Safari/537.36",
}
```
